Feature: Shopping List (everything)

rueblimaster opened the pull request:
## Dependencies
- #3562
- #3603


## What
Adds a general shopping list, can be added to by other features etc. also can break down items into their recipes

<details>
<summary>Images</summary>
![image](https://github.com/user-attachments/assets/60e09f20-8a1f-4b8a-aee8-a410f3519886)
</details>

## Changelog New Features
+ Added shopping list. - rueblimaster
    * Replaces Visitor Shopping List and craft material collector.

rueblimaster committed:
start with it

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
a little detekt

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    private fun MutableList<List<Any>>.drawShoppingList() {
+        if (itemList.isEmpty()) return
```
I know it's still draft but please use `listOf(Renderable)` instead of `listOf(List<Any>)`

rueblimaster replied:
ty, for your feedback and also no problem as else I would've probably left it like this.
I'm not gonna use this like this anymore anyways

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
add ShoppingListItem.toString()

+ continued a little

rueblimaster committed:
start with the rly basic logic etc stuff

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/config/commands/Commands.kt:
```diff
@@ -742,6 +743,11 @@ object Commands {
+        event.register("shtestshoppinglist") {
+            description = "Test the shopping list feature"
+            category = CommandCategory.DEVELOPER_TEST
+            callback { ShoppingList.test() }
```
Here

rueblimaster replied:
ty

rueblimaster committed:
figure out rendering

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
move command back here + some detekt

rueblimaster committed:
refractor NEU to Neu

rueblimaster committed:
added most of the add/remove logic

rueblimaster committed:
reduce amount of returns in 1 function

rueblimaster committed:
continue

rueblimaster committed:
fix warnings

rueblimaster committed:
some detekt

rueblimaster committed:
even more detekt

rueblimaster committed:
make the for loop good

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
add recipe getting

at least partially

rueblimaster committed:
detekt

rueblimaster committed:
add showSummary Config option

to implement

rueblimaster committed:
make big progress with the recipe etc

also add more commands

rueblimaster committed:
detekt

rueblimaster committed:
stop shadowing of it

detekt

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
you can now select recipes and other functionality

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
began on stopping recursive recipes

but only kinda works like I want it to, I don't want you to be able to show dia blocks as recipe for diamonds

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
make it work!!

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
laid out my plans

rueblimaster committed:
start with the plans

rueblimaster committed:
detekt

rueblimaster committed:
make categories hidable

and other stuff

rueblimaster committed:
Update ShoppingListItem.kt

rueblimaster committed:
kinda did a summary of all items needed

rueblimaster committed:
summary kinda thing done

rueblimaster committed:
implement pinning and fixed hiding

also added tooltips to items

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
start with storage

return side of things still needs work...

rueblimaster commented:
@Thunderblade73 
what do you think of the way I use the storage?

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,411 @@
+        }
+    }
+
+    fun saveToStorage() {
+        ProfileStorageData.profileSpecific?.shoppingList = Pair(categories.map { it.toEasyStorable() }, items.toEasyStorable())
+    }
+
+    fun loadFromStorage() {
+        storage?.let { (categories, items) ->
+            this.categories.clear()
+            this.categories.addAll(categories.map { ShoppingListCategory.fromEasyStorable(it) })
+            this.items.clear()
+            this.items.addAll(ShoppingListCategory.fromEasyStorable(items))
+        }
+    }
```
Those call should not exists. Since the storage can be used like normal so there is no need to translate. When you want to have a more space efficient layout in the json you would do a custom typeadatper.

rueblimaster replied:
yes sure, how do the custom typeadatpers work / where is an example for one?

Thunderblade73 commented on src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.kt:
```diff
@@ -776,4 +776,7 @@ class ProfileSpecificStorage {
     var enchantedClockBoosts: MutableMap<EnchantedClockHelper.SimpleBoostType, EnchantedClockHelper.Status> = enumMapOf()
+
+    @Expose
+    var shoppingList: Pair<List<Map<String, Any?>>, Map<String, Any?>> = Pair(listOf(), mapOf())
```
This is an unreadable mess. Pls use a Class and name the stuff properly. Also don't use Any.

hannibal002 replied:
i just moved the line for merge conflict fix, the change request is still valid (ignore the outdated tag)

Thunderblade73 reviewed (changes requested):
Only had a quick look at the storage code.

rueblimaster commented:
@Thunderblade73 
I'm thinking of saving past states of the shopping list for the user to undo their actions, I thought  about doing this via the json stuff as to reduce memory usages etc. What do you think about that?

hannibal002 committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.kt

hannibal002 committed:
merge conflict

hannibal002 commented:
Please use objects for clean and readable logic.
you can still store it as json, just make sure all fields of all objects have the expose annotation and are serializable (a data type we already store in config)

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,411 @@
+    fun loadFromStorage() {
+        storage?.let { (categories, items) ->
+            this.categories.clear()
+            this.categories.addAll(categories.map { ShoppingListCategory.fromEasyStorable(it) })
```
you confuse companion functions with extended functions here

rueblimaster replied:
yes, I'm not that much of a pro yet, but I'll figure it out (hopefully)

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListStorageUtils.kt:
```diff
@@ -0,0 +1,111 @@
+import at.hannibal2.skyhanni.utils.PrimitiveRecipe
+import net.minecraft.item.ItemStack
+
+fun ShoppingListCategory.toEasyStorable(): Map<String, Any?> {
```
please move all those functions into one object

hannibal002 replied:
Also since you create the ShoppingListCategory class, you can just move the extended functions inside this class directly

rueblimaster replied:
ye sure, I was just trying to abstract stuff into smaller files but in this case, that's not the way to go it seems

rueblimaster commented:
> Please use objects for clean and readable logic. you can still store it as json, just make sure all fields of all objects have the expose annotation and are serializable (a data type we already store in config)

so what do you think, should I scrap the undo stuff and just put the categories list and the items category into the storage?

hannibal002 commented:
Id say yes. the undo/redo stuff can be added later, and only over complicates it for now

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
idk why it's not working

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
still not working but idk why

rueblimaster commented:
do you have an idea why the storage isn't working?

rueblimaster committed:
Update ShoppingList.kt

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
fix it a little but still don't work?

rueblimaster committed:
redid storage, maybe now it works

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
made storage work!

rueblimaster committed:
reenable breakDownIntoSubitems

rueblimaster committed:
fixed not saving expanded recipes, replaced pinning

rueblimaster committed:
finish removing pinning

rueblimaster committed:
remove currentlyDecidingRecipe

rueblimaster committed:
implement left clicks

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
add Category option saveInStorage

rueblimaster committed:
start with own display item + other fixes

rueblimaster committed:
added possibly needed code

rueblimaster committed:
all the click actions work now

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
reworked clickable renderable stuff

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
continue with clicking

rueblimaster committed:
it's working!!!

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster commented:
oh shit, now I gotta do detekt

rueblimaster committed:
detekt

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListItem.kt:
```diff
@@ -0,0 +1,552 @@
+
+    fun fetchItemFromAvailableStorage() {
+        println(
+            "Fetching item from available storage: $internalName ${totalAmount.toInt()}, " +
+                +"getting ${getMissingAmountInInventory().toInt()}",
```
two plusses

rueblimaster replied:
you're right! didn't catch that, but it seems there was more...

Bloxigus commented on src/main/java/at/hannibal2/skyhanni/utils/renderables/Renderable.kt:
```diff
@@ -218,6 +246,116 @@ interface Renderable {
             }
         }
 
+        data class ClickTypeWithModifiers(
```
These overrides just reimplement the same behaviour as the default internal Kotlin implementations, is there any reason you are doing this?

Thunderblade73 replied:
The overrides are generally correct. But since this is declared as an data class the equals and hashCode function are redundantly declared here, since the data class already takes care of that.

rueblimaster replied:
yes, you're all right, I just implemented it like because it somehow fixed it overwriting in a map, but it's not needed anymore, is fixed

rueblimaster committed:
final detekt

rueblimaster committed:
improve ClickTypeWithModifiers

rueblimaster committed:
classify TODOs

add now, soon or later and maybe or not maybe

rueblimaster committed:
detekt ShoppingListCategory.kt

rueblimaster committed:
detekt Renderable.kt

rueblimaster committed:
as always: some detekt

rueblimaster committed:
detekt ShoppingList.kt

rueblimaster commented:
yay!

rueblimaster committed:
reduce print statements and add user feedback messages

Bloxigus commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,520 @@
+        saveShoppingList()
+    }
+
+    fun test() {
```
Could this be in a DebugDataCollectEvent handler?

rueblimaster replied:
sure, makes sense so why not
I hope it's expected like this

Bloxigus commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListCategory.kt:
```diff
@@ -0,0 +1,191 @@
+        ShoppingList.moveCategoryToTop(this)
+    }
+
+    fun onNormalRightClick() {
```
Are these logs necessary? Could these functions be moved to be inline with their only usages?

rueblimaster replied:
absolutely not, I forgot to change the system here

rueblimaster replied:
nope

rueblimaster committed:
Update ShoppingListItem.kt

rueblimaster committed:
switch category to new renderable

rueblimaster committed:
add category.isEmpty()

rueblimaster committed:
add onDebugDataCollect()

Bloxigus commented on src/main/java/at/hannibal2/skyhanni/data/SackApi.kt:
```diff
@@ -351,6 +351,7 @@ object SackApi {
         sackListNames = uniqueSackItems.map { it.itemNameWithoutColor.removeNonAscii().trim().uppercase() }.toSet()
     }
 
+    // TODO: also update sacks when items are added/removed from sack, not only on summary
```
You should probably remove TODOs if you aren't intending on doing it, and commented code that you don't intend on uncommenting

j10a1n15 replied:
have you seen the amount of todos skyhanni has? At this point it doesnt matter anymore
![image](https://github.com/user-attachments/assets/531983fe-32fe-40ce-ba21-9dfcea721af5)

hannibal002 replied:
This is a bad habit of mine. I would yell at myself not to add any more

hannibal002 replied:
Also most todos are about code design not missing features

rueblimaster replied:
removed commented out code

rueblimaster replied:
also I am kinda thinking about doing those TODOs, but this sadly this will be a eventually

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/CategoryTemplate.kt:
```diff
@@ -0,0 +1,34 @@
+
+import com.google.gson.annotations.Expose
+
+class CategoryTemplate {
```
I'm not gonna suggest the change on each of them individually, as there are a _lot_, but do not define explicit `constructor()`s for Kotlin classes where it isn't necessary to.

```suggestion
class CategoryTemplate(
	@Expose val name: String,
	@Expose val color: Char = '6',
	@Expose val hidden: Boolean = false,
	@Expose val items: List<ItemTemplate> = emptyList(),
) {
```

rueblimaster replied:
why? it just simplifies the code in the respective place it's called, doesn't it?

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ItemTemplate.kt:
```diff
@@ -0,0 +1,28 @@
+
+import com.google.gson.annotations.Expose
+
+class ItemTemplate {
```
Remove explicit constructor

rueblimaster replied:
I don't get how you expect me to do this? without having a new val that I don't want to expose

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/RecipeTemplate.kt:
```diff
@@ -0,0 +1,41 @@
+import at.hannibal2.skyhanni.utils.PrimitiveRecipe
+import com.google.gson.annotations.Expose
+
+class RecipeTemplate {
```
Remove explicit constructor

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,566 @@
+        }
+    }
+
+    @HandleEvent(onlyOnSkyblock = true, priority = HandleEvent.HIGH)
```
Doesn't seem this has a purpose to be high priority.

rueblimaster replied:
you're right, it just was like this where I looked for as reference

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,566 @@
+
+        clear()
+
+        add("aspect of the end".toInternalName(), 2.0, "Weapons")
```
What?

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,566 @@
+    }
+
+    // maybe name it removeCommand ???
+    fun remove(name: String, amount: Double? = null, categoryName: String? = null) {
```
This function is 3-4 nested if statements - refactor it into some helper functions if necessary, but this ain't it

rueblimaster replied:
here you go

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,566 @@
+
+        fun get(item: NeuInternalName) = allItems[item]
+    }
+
```
`.toSearchable()`

rueblimaster replied:
yes, what did I do wrong?

hannibal002 replied:
If the function does nothing else than all items(item) then you can inline the function entirely (idk the broader context where this needs searchable rn

DavidArthurCole commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,566 @@
+            val items = items
+
+            allItems.clear()
+            for (category in categories + items) {
```
Define some vals here instead of nestedly looping, especially using similar names, it gets unreadable quickly.

rueblimaster replied:
is this better like this?

rueblimaster committed:
reduce commented out code

rueblimaster committed:
small changes and better code

rueblimaster committed:
detekt

rueblimaster committed:
rewrite remove() for readability

rueblimaster committed:
item now uses internalName.createBuyTip()

rueblimaster committed:
detekt (once more...)

rueblimaster committed:
fixes

rueblimaster committed:
small fixes

rueblimaster committed:
implement more stuff

rueblimaster committed:
migrate visitor shopping list

rueblimaster committed:
remove unused code

rueblimaster committed:
fix stuff

also remove some commented out code

hannibal002 commented:
> ### 1 Detekt Failure was detected:
> 
>     * [ShoppingListItem.kt#L367](https://github.com/hannibal002/SkyHanni/blob/89c223d63f227bf66003ec1a79fd70d3ca4405b6/src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListItem.kt#L367): The function getRenderables appears to be too complex based on Cyclomatic Complexity (complexity: 25). Defined complexity threshold for methods is set to '25'

This is a intended check, please make that function shorter by splitting it into different functions or compact the logic if possible

rueblimaster commented:
> > ### 1 Detekt Failure was detected:
> > ```
> > * [ShoppingListItem.kt#L367](https://github.com/hannibal002/SkyHanni/blob/89c223d63f227bf66003ec1a79fd70d3ca4405b6/src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListItem.kt#L367): The function getRenderables appears to be too complex based on Cyclomatic Complexity (complexity: 25). Defined complexity threshold for methods is set to '25'
> > ```
> 
> This is a intended check, please make that function shorter by splitting it into different functions or compact the logic if possible

yes sure, I'm doing this rn actually, also this make so much sense, it's absolutely a nightmare rn lol

rueblimaster committed:
implement middle click

also split getRenderables()

rueblimaster committed:
reduce secondary constructors

rueblimaster committed:
detekt

rueblimaster committed:
moved all template conversion logic into templates

rueblimaster commented:
@DavidArthurCole @hannibal002 
I hope I did the right thing with the constructors, is this good like this?

rueblimaster committed:
search? not working...

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,610 @@
+            inventoryOpen = true
+            update()
+        }
+        config.position.renderRenderables(listOf(display.buildSearchBox(SearchTextInput())), posLabel = "Shopping List")
```
```suggestion
        config.position.renderRenderable(display.buildSearchBox(SearchTextInput()), posLabel = "Shopping List")
```

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,610 @@
+    @HandleEvent(onlyOnSkyblock = true)
+    fun onRender(event: GuiRenderEvent.GuiOverlayRenderEvent) {
+        if (!isEnabled()) return
+        config.position.renderRenderables(listOf(display.buildSearchBox(SearchTextInput())), posLabel = "Shopping List")
```
```suggestion
        config.position.renderRenderable(display.buildSearchBox(SearchTextInput()), posLabel = "Shopping List")
```

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -0,0 +1,610 @@
+import at.hannibal2.skyhanni.utils.PrimitiveItemStack.Companion.toPrimitiveStackOrNull
+import at.hannibal2.skyhanni.utils.PrimitiveRecipe
+import at.hannibal2.skyhanni.utils.RecipeType
+import at.hannibal2.skyhanni.utils.RenderUtils.renderRenderables
```
```suggestion
import at.hannibal2.skyhanni.utils.RenderUtils.renderRenderable
```

rueblimaster committed:
Merge branch 'hannibal002:beta' into shopping-list

rueblimaster committed:
little oopsie

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

rueblimaster committed:
not much

rueblimaster committed:
improve test() contents

rueblimaster committed:
cache interalNames

rueblimaster committed:
small fixes in item

rueblimaster committed:
implement arrow up + down and other small improvements

rueblimaster committed:
make rendering stuff better

rueblimaster committed:
own detekt

rueblimaster committed:
detekt

rueblimaster commented:
> ### 1 Detekt Failure was detected:
> 
>     * [ShoppingListItem.kt#L399](https://github.com/hannibal002/SkyHanni/blob/bd7b6e560fbf18e2754af0898d0fd82183360c6a/src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingListItem.kt#L399): The function getClickLayout is too long (115). The maximum length is 100.

meh, nah

rueblimaster committed:
massively improve commands

rueblimaster committed:
remove done TODO

hannibal002 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -103,6 +103,44 @@ object ShoppingList {
 
     // all the functions for interacting with the shopping list come here
+    // parseCommandArguments only works if only the item name is also okay
+    fun parseCommandArguments(args: Array<String>): Triple<String, Double?, String?>? {
```
`Triple<String, Double?, String?>`  please  create a new object instead of using a triple

rueblimaster committed:
create custom returned class CommandArguments

rueblimaster committed:
detekt

rueblimaster commented:
gotta replace bazaar buy recipe ingredients helper

j10a1n15 commented on src/main/java/at/hannibal2/skyhanni/features/inventory/shoppinglist/ShoppingList.kt:
```diff
@@ -102,9 +102,11 @@ object ShoppingList {
     var currentlyOpenRecipe: PrimitiveRecipe? = null
     var displayItem: ItemStack? = null
 
+    class CommandArguments(val itemName: String, val amount: Double?, val categoryName: String?)
```
either make this private or name it better
having a publicly accessable class named CommandArguments that only exists for shoppinglist will confuse

j10a1n15 commented:
> gotta replace bazaar buy recipe ingredients helper

yes please make the pr even bigger

rueblimaster commented:
> > gotta replace bazaar buy recipe ingredients helper
> 
> 
> 
> yes please make the pr even bigger

I mean we can also have duplicate features if you'd like to...
But what I need to do for it is really only remove the bazaar one

rueblimaster committed:
make CommandArguments private

rueblimaster committed:
detekt

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/visitor/GardenVisitorFeatures.kt

rueblimaster committed:
remove CraftMaterialCollector

rueblimaster committed:
little improvement

rueblimaster committed:
make ItemsOverall more readable

rueblimaster committed:
make config option actually work...

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
fix merge

rueblimaster committed:
really fix merge (hopefully)

rueblimaster committed:
fix it?

rueblimaster committed:
I present you: the correct fix

and some minor code improvements

rueblimaster committed:
Merge pull request #3 from rueblimaster/fix-visitor-shopping-list-falsely-hiding

Fix: Visitor Shopping List disappearing while supercrafting

rueblimaster committed:
Revert "Fix: Visitor Shopping List disappearing while supercrafting"

rueblimaster committed:
Merge pull request #4 from rueblimaster/revert-3-fix-visitor-shopping-list-falsely-hiding

Revert "Fix: Visitor Shopping List disappearing while supercrafting"

rueblimaster committed:
finally: Fix: Visitor Shopping List disappearing while supercrafting

and other visitor shopping list fix

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/visitor/GardenVisitorFeatures.kt

CalMWolfs committed:
fix merge

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/features/garden/visitor/GardenVisitorFeatures.kt

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
add config options

to implement

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/utils/InventoryUtils.kt

CalMWolfs committed:
Merge branch 'refs/heads/beta' into fork/rueblimaster/shopping-list

# Conflicts:
#	src/main/java/at/hannibal2/skyhanni/utils/InventoryUtils.kt

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
remove unneeded code change

rueblimaster committed:
remove unneeded TODo

rueblimaster committed:
revert searchable

rueblimaster committed:
Update ShoppingList.kt

rueblimaster committed:
detekt remove import

rueblimaster committed:
add middle mouse

rueblimaster committed:
add PrimitiveItemStack.toPrimitiveIngredient

hannibal002 commented:
You wanna split the 1.8k line pr into 5 prs all around 300 lines, not into 50 prs all around 30 lines.please thank you

rueblimaster commented:
> You wanna split the 1.8k line pr into 5 prs all around 300 lines, not into 50 prs all around 30 lines.please thank you

yes, to reduce the size of the main logic part I wanted to get rid of some smaller changed files

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
merging and sanatizing

rueblimaster committed:
Merge branch 'shopping-list' of https://github.com/rueblimaster/SkyHanni into shopping-list

rueblimaster commented:
sry, I broke it or smth

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
remove redundant default values from CategoryTemplate

rueblimaster committed:
formatting

rueblimaster committed:
Merge remote-tracking branch 'upstream/beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
fix import

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
fix merge creating empty file

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
imports

rueblimaster committed:
detekt

rueblimaster committed:
updates

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Delete CraftMaterialCollector.kt

rueblimaster committed:
Make position property immutable in ShoppingListConfig

Changed the 'position' property from 'var' to 'val' in ShoppingListConfig to make it immutable. This enhances safety by preventing unintended modifications to the position.

rueblimaster committed:
Fix concurrent modification in visitor item removal

Convert category.items to a list before iteration to avoid concurrent modification exceptions when removing items during iteration.

rueblimaster committed:
Refactor ShoppingList rendering logic

Moved rendering logic into a new render() function and updated event handlers to use it. Added handling for ScreenDrawnEvent to support rendering in GuiEditSign. Improved type safety for config property.

rueblimaster committed:
Add RecipeInventory utility for tracking open recipes

Introduces the RecipeInventory object to track currently open crafting recipes in the inventory. Handles events for opening and closing recipe inventories, extracting ingredients and output, and storing the current recipe state.

rueblimaster committed:
Add RecipeResolver utility for recipe selection

Introduces the RecipeResolver class to handle resolving and selecting crafting recipes for items, including user interaction for multiple recipe options and event handling for GUI integration.

rueblimaster committed:
Specify nullable type for output property

Changed the type of the 'output' property in PrimitiveRecipe to explicitly be 'PrimitiveIngredient?' to clarify its potential nullability.

rueblimaster committed:
Refactor recipe handling in ShoppingListItem

Replaces direct recipe management with a RecipeResolver class for improved encapsulation and deferred recipe resolution. Removes redundant recipe selection logic and simplifies subitem breakdown, delegating recipe decision and error handling to the new resolver.

rueblimaster committed:
Finish refactoring recipe handling in shopping list feature

Moved currentlyOpenRecipe management to RecipeInventory and updated related logic in ShoppingList. Made RecipeResolver's recipe property private and added a setRecipe method for controlled updates. Removed unused item click handling from ShoppingListCategory.

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Refactor shopping list item subitem handling

Replaces the single subItems list in ShoppingListItem with displayedSubItems and backgroundSubItems to better separate UI and background logic. Updates related methods and serialization to use the new structure, and adds @Expose annotations for Gson compatibility. Also improves recipe resolution and ingredient handling for more accurate item breakdowns.

rueblimaster committed:
detekt

rueblimaster committed:
Add isCraftable method and update item status logic

Introduced the isCraftable() method to determine if a shopping list item can be crafted based on available subitems. Updated status indicators and click actions to use isCraftable, improving clarity between having all items and being able to craft.

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Add Shopping List config to InventoryConfig

Introduces a new ShoppingListConfig property to InventoryConfig, categorized as 'Shopping List' for general shopping list features.

rueblimaster committed:
Fix clickable text rendering in Renderable

Replaces the use of string(text) with text(text) in the clickable function to ensure proper rendering of the text parameter.

rueblimaster committed:
Refactor ingredient count calculation and cleanup

Simplifies the logic for updating ingredient counts by using the safe call and elvis operator. Removes an unnecessary cast and an unused import for cleaner code.

rueblimaster committed:
Merge branch 'beta' into shopping-list

rueblimaster committed:
Merge branch 'beta' into shopping-list
